### ISSUE TITLE
.rubocop.yml exclude doesn't work.

### ISSUE THREAD

**Issue by ghost** [rubyide/vscode-ruby#227]
### Your environment

- `vscode-ruby` version: 0.15.0
- Ruby version: ruby 2.4.2p198 (2017-09-14 revision 59899) [x64-mingw32]
- VS Code version: 1.18.1 and 1.19.0-insider 
- Operating System: Windows 10 latest

### Expected behavior

`.rubocop.yml` like this should work:

```
AllCops:
  Exclude:
    - 'db/**'
    - '**/Gemfile'
  TargetRubyVersion: 2.4
```

### Actual behavior

`db/schema.rb` and `Gemfile` being rubocopped.

Via `rubocop db` it works. So, **it's a bug**?

**Comment by karuppasamy** [comment]
I don't have any experience with Windows 10. But, I am guessing the below config (in user settings) will solve your problem. 

> "ruby.rubocop.configFilePath": "path/to/your/.rubocop.yml",

**Comment by johnunclesam** [comment]
I can't because today is not possible: https://github.com/rubyide/vscode-ruby/issues/133

**Comment by johnunclesam** [comment]
But anyway other settings in my rubocop.yml work good!

**Comment by wingrunr21** [comment]
Hi @johnunclesam 

Is this still an issue for you with the latest version of the extension? Some changes were introduced in how Rubocop is run in v0.16.0 so wondering if that fixed it.

**Comment by WA9ACE** [comment]
I personally still have this problem and I'm on 0.17.0

**Comment by msokk** [comment]
Still a problem for me, v0.17.0 with rubocop v0.53.0 on macOS. `schema.rb` still gets linted by extension.

**Comment by johnunclesam** [comment]
Still a problem. `schema.rb` still gets linted by extension.

**Comment by wingrunr21** [comment]
Closing for issue cleanup. Apologies if this is still an issue. We are working to improve the core extension experience.

**Comment by gurgeous** [comment]
I bet rubocop doesn't apply `Exclude` rules when you run it on a single file. Perhaps this will help - https://github.com/bbatsov/rubocop/issues/893

**Comment by garyking** [comment]
I just ran Rubocop on a single file, and it respected exclusions. 

Specifically, certain rules by default in Rubocop, don't apply to files in `specs` and `test` folders, yet in this extension, they still run.

**Comment by gurgeous** [comment]
@garyking, there are two separate issues:

1. Things like `Metrics/ClassLength` are applied when they shouldn't be. Presumably because rubocop isn't getting enough context when linting. Haven't investigated yet.
2. Files that are listed under `Exclude` still get linted (like `db/schema.rb`). I bet adding `--force-exclusion` will solve this problem.

**Comment by wingrunr21** [comment]
@garyking what's your rubocop config for this extension look like? Both of those issues basically mean Rubocop is either not finding the `.rubocop.yml` or is ignoring the settings. I'm wondering if one of the extension settings is activating a CLI option that causes things to be ignored.

**Comment by garyking** [comment]
This ext isn't importing Excludes from inherited rules for some reason.

I have a workspace in VSC. There's a `.rubocop.yml` in the workspace's root. It just has: `inherit_from: /Users/gary/.base-rubocop.yml`.

In that inherited file, it has:

```yaml
Metrics/ClassLength:
  Exclude:
    - 'test/**/*'
  Max: 150
```

Back in my workspace, in the file `test/test_file.rb`, if I run Rubocop on the file from Terminal, then I get no errors. But in VSC, this ext gives me the `Metrics/ClassLength` error, indicating that `Exclude` is not inherited. But the `Max` value definitely is, because it says `205/150`.

**Comment by gurgeous** [comment]
I am planning to investigate this and other linting issues, see https://github.com/rubyide/vscode-ruby/issues/317#issuecomment-380245607. Probably need a new tracking issue for that once @wingrunr21 chimes in.

**Comment by zackchandler** [comment]
Changing:

```yaml
inherit_from: .rubocop_todo.yml
```

to

```yaml
inherit_from: ~/.rubocop_todo.yml
```

fixed a similar issue for me.

**Comment by mpressen** [comment]
exclude in .rubocop.yml when using vs code formatting on save still doesn't work.

**Comment by wingrunr21** [comment]
It will be fixed soon

**Comment by cliffrowley** [comment]
Any news on this?

**Comment by wingrunr21** [comment]
The language server support for RuboCop should support the whole config file

**Comment by cliffrowley** [comment]
I am still getting warnings in files that I have specifically excluded on my `.rubucop.yml`.

**Comment by mattalat** [comment]
I want to follow up on this and say that the following configuration will not properly exclude some `folder/file.rb`

```yml
AllCops:
  Exclude:
    - folder/**/*
    - folder/*
    - folder/*.rb
```

However, explicitly supplying the path does seem to work for me:

```yml
AllCops:
  Exclude:
    - folder/file.rb
```

**Comment by mattalat** [comment]
I'll chime in and say that the AllCops ignore doesn't seem to have any effect, but other settings in the `.rubocop.yml` file are respected by the extension.

**Comment by corselia** [comment]
The `Exclude:` settings in my `.rubocop.yml` doesn't work, too.
It's tough for me.

**Comment by jacek213** [comment]
Same here, not cool at all.

**Comment by nemoDreamer** [comment]
Yup, still an issue with a config file that inherits from other files...

**Comment by wingrunr21** [comment]
Can I get some more examples of RuboCop configs? I'm wondering if this is a confluence of three things:

- The file name is passed into RuboCop's `stdin` input as an absolute path
- RuboCop configs are probably configured as relative paths
- The working directory RuboCop is run against

Note that other than how the language server invokes RuboCop, there's little that I can do here. I need to be able to invoke RuboCop consistently and have it give me an appropriate response in whatever circumstance. This is why I think there's some kind of mismatch between how everyone's configuring their linters and how RuboCop is being invoked.

**Comment by pre** [comment]
When path to `inherit_from` is relative, vscode does not find that file. 

For example in a monorepo you could have `inherit_from: ../../common/rubocop-defaults.yml`

This works from the command line as the current working directory can explicitly be defined. But even though I start VSCode in that same CWD, Rubocop execubted by VSCode does not find that mentioned file in the relative path.

**Comment by pre** [comment]
This issue was closed so I explained my findings in https://github.com/rubyide/vscode-ruby/issues/566#issuecomment-610258250

TL;DR

* `inherit_from` works with relative paths for me (when `code` is executed in the project's root folder) 
* relative paths in the `rubocop.yml` referred by `inherit_from` do not work as the Current Working Directory is not respected

**Comment by dansteele** [comment]
I get this issue too. It's wonderful most of the time but in those 1% of files where it's not folloing the .yml it becomes a pain and I can't format on save

@wingrunr21 [here's my .rubocop.yml](https://gist.github.com/dansteele/1142c3700fb9860ec7b3defccbaf2c4f).

**Comment by mmirus** [comment]
:wave: I'm running into this as well!

`.rubocop.yml`

```rb
inherit_gem:
  rubocop-shopify: rubocop.yml

AllCops:
  TargetRubyVersion: 2.5
  Exclude:
    - 'spec/**/*'
    - 'db/migrate/**/*'

```

**Comment by thedanielvogelsang** [comment]
?

**Comment by jensljungblad** [comment]
Yep this isn't working for me either. Not using any inheritance, just a plain Rubocop file with

```
AllCops:
  Exclude:
    - bin/*
    - db/schema.rb
    - node_modules/**/*
    - vendor/**/*
```

And using the default plugin settings:
```
"ruby.codeCompletion": "rcodetools",
    "ruby.format": "rubocop",
    "ruby.intellisense": "rubyLocate",
    "ruby.useLanguageServer": true,
    "ruby.useBundler": true,
    "ruby.lint": {
        "rubocop": {
            "useBundler": true
        }
    }
```

**Comment by danielsvane** [comment]
You can use:
```
"ruby.lint": {
    "rubocop": {
        "useBundler": true,
        "forceExclusion": true
    }
}
```

**Comment by jensljungblad** [comment]
> You can use:
> 
> ```
> "ruby.lint": {
>     "rubocop": {
>         "useBundler": true,
>         "forceExclusion": true
>     }
> }
> ```
> 
> The issue arises because the extension runs `rubocop` on a single file, which doesn't use the exclude settings from rubocop.yml unless you run it with `--force-exclusion`.

That fixed it, thanks!

**Comment by mmirus** [comment]
Fixed it for me, as well! Didn't need `"useBundler": true`. Thank you!

**Comment by 907th** [comment]
Still doesn't work for me (v0.28.1). Even with `forceExclusion: true` option.

Non-working rule:
```
Metrics/BlockLength:
  Max: 30
  Exclude:
    - "spec/**/*.rb"
    - "config/routes/*.rb"
    - "engines/member_api/config/routes.rb"
    - "engines/admin_api/config/routes.rb"
```

Run log:
```
Lint: executing bundle exec rubocop -s '/Users/laise/Work/firm/proj/spec/services/ticket_service/file_uploader_spec.rb' -f json --force-exclusion...
```

Error sown in editor: `Block has too many lines. [62/30]Metrics/BlockLength(RuboCop)`

**Comment by jensljungblad** [comment]
Yep, stopped working for me again, even with `forceExclusion`...

**Comment by FooBarWidget** [comment]
I think I found the cause of this bug. See #717.

**Comment by ssoulless** [comment]
In my .rubocop.yml I have this

```
Metrics/BlockLength:
  IgnoredMethods: ["describe", "context"]

```

And in my VScode settings Json I have this 

```
"ruby.lint": {
        "rubocop": {
            "forceExclusion": true
        }
    }
```

it still shows the `Block has too many lines. Metrics/BlockLength(RuboCop)` in the editor.

**Comment by jensljungblad** [comment]
I happened to notice something regarding this. The `forceExclusion: true` actually does work if I also turn off `ruby.useLanguageServer`. Seems like the combination of those two options (for me at least) causes the problem.

**Comment by mmirus** [comment]
The `forceExclusion` option is no longer solving this for me, unfortunately. Disabling `ruby.useLanguageServer` does stop the problem, but that's obviously not a good solution.

**Comment by fedeth** [comment]
The `forceExclusion` option combined with` "useBundler": true` works for me.

**Comment by peternixey** [comment]
I had the same problem but noticed that rubocop wasn't running from the same root directory as my project. So I tried making the exclusion root-independent and it then worked fine:

```yaml
# Doesn't work :(
AllCops:
  Exclude:
    - db/schema.rb
```
was failing because `db` was not in the same directory that Rubocop was looking for it.

```yaml
# Works :)
AllCops:
  Exclude:
    - */schema.rb
```

**Comment by heragu** [comment]
@peternixey I tried your solution and works in vs code but then it does not work in the command line. When I tried to run in the terminal `bundle exec rubocop` I got an error like the following:

```bash
/path/to/your/project/.rubocop.yml): did not find expected alphabetic or numeric character while scanning an alias at line 5 column 7
/path/to/your/home/.rbenv/versions/2.7.1/lib/ruby/2.7.0/psych.rb:456:in `parse'
/path/to/your/home/.rbenv/versions/2.7.1/lib/ruby/2.7.0/psych.rb:456:in `parse_stream'
/path/to/your/home/.rbenv/versions/2.7.1/lib/ruby/2.7.0/psych.rb:390:in `parse'
...
```

This helps while developing with VS code but if you have `rubocop` integrated in your deployment pipeline it is not a final solution.

**Comment by peternixey** [comment]
@heragu - it actually doesn’t even work in Vscode any more for me. I find anything to do with Ruby (solargraph etc too) a bit of a black art to get working!

**Comment by evgenyneu** [comment]
The `"forceExclusion": true` solution does not work for me. What's the status of this bug? Is anyone working on it?

**Comment by shrkw** [comment]
This behavior is really strange. I created experiment project: https://github.com/shrkw/vscode-ruby-dev-env-sample , and rubocop and "Exclude" settings works very well yesterday.
I opened it today, "Exclude" settings does not work at all!

<img width="264" alt="スクリーンショット 2022-04-19 16 10 04" src="https://user-images.githubusercontent.com/275147/163946813-9c20ad8d-9cf1-4b3b-b5ca-ee6645fc304a.png">

Debug Output view for "Ruby Language Server" shows rubocop has correct option...

```
Lint: executing bundle exec rubocop -s '/Users/shirakawa2/Development/sandbox-ruby-vscode/src/apple.rb' -f json --force-exclusion...
Lint: executing bundle exec rubocop -s '/Users/shirakawa2/Development/sandbox-ruby-vscode/src_excl/cherry.rb' -f json --force-exclusion...
```

and same command returns no offenses.

``` $ bundle exec rubocop  '/Users/shirakawa2/Development/sandbox-ruby-vscode/src_excl/cherry.rb' -f json --force-exclusion
{"metadata":{"rubocop_version":"1.27.0","ruby_engine":"ruby","ruby_version":"2.6.8","ruby_patchlevel":"205","ruby_platform":"universal.arm64e-darwin21"},"files":[],"summary":{"offense_count":0,"target_file_count":0,"inspected_file_count":0}}%      
```

It's really difficult to understand.

**Comment by benzado** [comment]
Since this is an old issue with lots of comments, it's possible there are multiple bugs being represented here. However, I strongly suspect this issue is a duplicate of #578 and #719, and is fixed by PR #720. However, the fix hasn't made it into a release yet.

**Comment by nruth** [comment]
> I found what caused the problem, it's this setting:
> 
> ```
> "ruby.lint": {
>    "rubocop": {
>       "useBundler": true // enable rubocop via bundler
>    }
> }
> ```
> 
> I removed it and it worked! It should be `ruby.rubocop.useBundler` and NOT `ruby.lint.rubocop.useBundler`.

this works well, but looks worrying because vscode greys it out and says the config is unknown. Where would that need updating?

**Comment by qwell** [comment]
Even if PR #720 gets merged and doesn't work, we're no worse off than we are right now with 0.28.1.  Could #641 even be reproduced at this point, given the 78 (I counted) releases rubocop has had since August 01, 2020?

Rather than dealing with the fallout of that commit for 2 more years, can PR #647 just be reverted until it can be shown that there actually is a problem in the first place?

**Comment by qwell** [comment]
#840 (which should probably supersede #825 - although @senhalil should feel free to merge this and let things continue over there) finishes reverting all of the changes from #647, which hadn't already been reverted elsewhere.

**Comment by celesteking** [comment]
You should really also document the pattern syntax. Whether it's a `Dir.glob` or `fnmatch`, or shell glob, explicitly state so. If it's a custom syntax, document it.
